--- curl.orig/Makefile.am	2026-05-30 19:49:02.000000000 +0800
+++ curl/Makefile.am	2026-05-30 19:49:04.000000000 +0800
@@ -83,8 +83,8 @@ DISTCLEANFILES = buildinfo.txt
 
 bin_SCRIPTS = curl-config
 
-SUBDIRS = lib docs src scripts
-DIST_SUBDIRS = $(SUBDIRS) tests packages include docs
+SUBDIRS = lib src
+DIST_SUBDIRS = $(SUBDIRS) packages include
 
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = libcurl.pc
@@ -180,12 +180,9 @@ rpm:
 if BUILD_DOCS
 install-data-hook:
 	(cd include && $(MAKE) install)
-	(cd docs && $(MAKE) install)
-	(cd docs/libcurl && $(MAKE) install)
 else
 install-data-hook:
 	(cd include && $(MAKE) install)
-	(cd docs && $(MAKE) install)
 endif
 
 # We extend the standard uninstall with a custom hook:
--- curl.orig/Makefile.in	2026-05-30 19:49:02.000000000 +0800
+++ curl/Makefile.in	2026-05-30 19:49:04.000000000 +0800
@@ -498,8 +498,8 @@ EXTRA_DIST = CHANGES.md COPYING RELEASE-
 
 DISTCLEANFILES = buildinfo.txt
 bin_SCRIPTS = curl-config
-SUBDIRS = lib docs src scripts
-DIST_SUBDIRS = $(SUBDIRS) tests packages include docs
+SUBDIRS = lib src
+DIST_SUBDIRS = $(SUBDIRS) packages include
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = libcurl.pc
 all: all-recursive
@@ -1134,11 +1134,8 @@ rpm:
 # We extend the standard install with a custom hook:
 @BUILD_DOCS_TRUE@install-data-hook:
 @BUILD_DOCS_TRUE@	(cd include && $(MAKE) install)
-@BUILD_DOCS_TRUE@	(cd docs && $(MAKE) install)
-@BUILD_DOCS_TRUE@	(cd docs/libcurl && $(MAKE) install)
 @BUILD_DOCS_FALSE@install-data-hook:
 @BUILD_DOCS_FALSE@	(cd include && $(MAKE) install)
-@BUILD_DOCS_FALSE@	(cd docs && $(MAKE) install)
 
 # We extend the standard uninstall with a custom hook:
 uninstall-hook:
